File: cat_boxing/caged_cat/c++/caged_cat.cpp
/*-------------caged_cat.cpp--------------------------------------------------//
*
* Purpose: To place cats in cages and place those cages in a truck in the most
*          Efficient way possible. 
*
*-----------------------------------------------------------------------------*/

#include <iostream>
#include "fluid_cat.h"

// See that line? The one with the "fluid_cat.h" just above? Yeah, that's a 
// little important. In order to link files from one code to another, you use
// a "header" file (thus the ".h" extension). It is certainly worthwhile to look
// into that file and see what you can learn!

using namespace std;

// the other two lines should make sense. We are including a module to allow us
// to stream output to the terminal and we are using standard notation

// Now comes a somewhat challenging part, planning out the code. In truth, this
// shouldn't be too bad. We just need to generate a random cat, place that cat
// in an appropriately sized box, and place that box in a truck. The third step
// might be a bit of a challenge, though.

// So let's break it down step by step.
// First function: box_cat. This function will take a random cat size and place
// it into the appropriately sized cage. In scientific applications, this 
// process is known as "binning" data, because you are taking data and placing
// it in bins (or in this case, you are placing cats in cages). It is a very 
// useful tool to know how to use!

// This function is not too bad. It will take in a cat size and spit out an 
// integer value (either 1, 2, or 3) that designates the size of the cage. 
// for the purposes of this approximation, small boxes are 1/3 the size of large
// boxes. The medium sized boxes are 2/3 the size of the large boxes. 

int box_cat(){

    int cage_size;
    // First, we need to create an integer variable for the size of the cages.
    // Now, let's call the function from our previous example: the fluid cat 
    // approximation. What function do we need? Well, the one that generates
    // random cats, of course! This is what the header file is for, calling
    // functions from other files and using them whenever you want.

    int cat_size = generate_cat();

    // Cool, we have a cat size. Now all we need to do is place them into boxes.
    // How would we do that in real life? Simple! If the cat is large, we place
    // it in a large box. If it is small, we place it in a small box. If it is
    // a moderate size, we place it in a moderately sized box. Luckily for us, 
    // c++ (and most other languages) understand the word "if." While coding, an
    // "if" statement tells the computer to do something when certain condition
    // are met. 

    // If I remember correctly, we set our car_size to be in a range from 1 to 
    // 100. Now that I think about it, this is not a plasible scale... but who
    // cares? If our cats are between 1 and 100, then we have 3 "size ranges"
    // to worry about: cat_size <= 33, 33 < catsize <= 66, and 
    // 66 < cat_size <= 100. Note that the "<=" means "less than or equal to."
    // It is similar to every other comparative sign, like "<", ">", "==", and
    // ">=". We just need to cover the full range between 1 and 100.

    if (cat_size <= 33){
        cage_size = 1;
    }

    // look carefully at this syntax. The squirrelly brackets "{}" open and
    // close the statement, while the parentheses designate the condition 
    // necessary to go into the loop.

    // Now let's do this twice more!

    if (cat_size > 33 && cat_size <= 66){
        cage_size = 2;
    }

    // Check that "if" statement out! Multiple conditions? That's what I call
    // a good time!

    if (cat_size > 66 && cat_size <= 100){
        cage_size = 3;
    }

    // Now all we need to do is return the cage_size and feed that into the next
    // section: placing it in the truck. *dramatic music*

    return cage_size;
}

// Let's think about this function. We want to place boxes of certain sizes into
// a truck. It doesn't sound bad, right? Well... it can be. There are an awful
// large number of configurations that are simply inefficient. I guess the best
// bet for our first attempt at filling the truck is to parse the inside of the
// truck into three columns: one for large, medium, and small boxes. In this 
// case, we can assume that the truck can hold 4 large boxes, 6 medium boxes, 
// and 12 small boxes. 

// What happens with the extra boxes? Well, we'll just throw them to the side
// and leave them for the next truck. I think it would be interesting to count 
// the number of caged cats that we will be leaving at the side of the road,
// though. So let's do that too.

int organize_cats(int& is_full, int& large_count, 
                  int& med_count, int& small_count){

    // I have made this function dependent on the variable "is_full", which is
    // just a query of whether the truck is full or not. If is_full reads the
    // number "3", it is full of small, medium, and large cages. I have also 
    // made the function dependent on the large, medium, and small counts to
    // keep track of how many cats have passed through our robot's claws.

    // But wait! What's up with the "&" character at the end of the "int" 
    // arguments? Well, I'm glad you asked! Those are notifiers to the compiler
    // to let it know that it does not have to recreate those variables. They 
    // are somewhere else in the code. This way, we can return all of those
    // integer values at the end. If this is confusing, don't worry! It'll be
    // easier to understand with time!

    // So, first things first, let's establish the appropriate variables.

    int large_num = 4, med_num = 6, small_num = 12;

    // Now, let's place a cat in a cage.

    int cage_size = box_cat();

    // There are three possibilities: 
    //     The cage is large (3) 
    //     The cage is medium-sized (2)
    //     The cage is small (1)
    // We need an "if" statement to deal with all three cases. But what do we do
    // after that? Well, the first "if" statement will be for us to establish
    // the box size in this function. We then need to place it in the 
    // appropriate position. In this case, though, that doesn't really matter. 
    // We just need to know that we can fit 4 large, 6 medium-sized, and 12 
    // small boxes in the truck. If we have all that can fit of a particular
    // type of box, we also need to pass that information along. We'll use a 
    // fourth "if" statement for this.

    if (cage_size == 3){

        // This corresponds to large boxes. Let's first update the count.

        large_count = large_count + 1; 

        // now we need an additional statement to let us know when the part of
        // truck designated to store large boxes is full

        if (large_count == 4){

            is_full = is_full + 1;

        }
    }

    // Now we need to do it twice more for the other cage sizes

    if (cage_size == 2){

        med_count = med_count + 1;

        if (med_count == 6){
            is_full = is_full + 1;
        }

    }

    if (cage_size == 1){

        small_count++;

        if (small_count == 12){
            is_full++;
        }
    }

    // Now we just need to return our values!

    return is_full, small_count, large_count, med_count;
}

// Now let's start the main function!

int main(){


    // First, let's create the variable to check whether the truck is full.

    int is_full = 0, large_count = 0, med_count = 0, small_count = 0;

    // Now let's create a loop to keep perfomring calculations until the truck
    // is full.

    while (is_full < 3){

        organize_cats(is_full, large_count, med_count, small_count);

    }

    cout << "The truck is full! Let's move out and leave "
         << large_count - 4 << " large cats, "
         << med_count - 6 << " medium-sized cats, and "
         << small_count - 12 << " small cats behind! Who needs 'em?"
         << endl;
}

File: cat_boxing/caged_cat/python/caged_cat.py
#-------------caged_cat.py-----------------------------------------------------#
#
# Purpose: To give te cats we previously worked on cages and then stuff them 
#          into a truck. This is a step up from last time where we just assumed
#          the cats were fluid and shoved them into a box. We will also keep
#          track of all the cats we must leave on the side of the road to fill
#          the truck optimally (leaving some cats for the next truck).
#
# Assumptions: 
#    1. 1 cat per cage, 3 cage sizes
#    2. The truck can hold 4 large, 6 medium-sized, and 12 small cages
#    3. The truck will be filled optimally. Any cats left over will wait for
#       the next truck. 
#
# I think that's it! So let's get to it!
#------------------------------------------------------------------------------#

# First things, first. We are going to need to generate a randomly sized cat 
# again, so let's import our last file, which has been copied to this directory.

from fluid_cat_slim import *

# Note the syntax. The "*" symbol means that we are importing all the functions
# from the file "fluid_cat_slim.py". Also note that the ".py" extension is no
# present. That might trip you up in the future.

# Like last time, we'll start function by function. We'll box the cats and place
# them in the truck

def box_cat():

    # Remember the "generate_cat" function from last time. We need that.

    cat_size = generate_cat()

    # Now we have a cat of some given size. Let's place that cat in a box.

    if cat_size <= 33:
        cage_size = 1

    # Did you see that? That was an "if statement"! It works as you would 
    # expect. When the condition is met, the action in the statement will
    # be performed. Simple as that. In this case, "cage_size" will be defined as
    # 1. That is all.

    if cat_size > 33 and cat_size <= 66:
        cage_size = 2

    # Here we are showing off an "if statement" with mutliple conditions

    if cat_size > 66 and cat_size <=100:
        cage_size = 3

    # Honestly, we didn't need the upper limit of 100, but it's in there for
    # clarity's sake.

    # Alright! one function done! now all we need to do is return cage_size
    return cage_size

# Now, let's think about this upcoming function a bit. We want to count all the
# left over cats, so we will need a counter for each of those. On top of that,
# we will need a counter to let us knoe when the truck is full of large, 
# medium-sized, and small boxes. This means that this function should be 
# dependent on those 4 variables.

def organize_cats(is_full, large_count, med_count, small_count):

    # Let's go ahead and define the maximum number of large, medium-sized, and 
    # small boxes our truck can hold

    large_num = 4 
    med_num = 6
    small_num = 12

    # Now let's place a cat in a cage

    cage_size = box_cat()

    # There are 3 possibilities, at this point:
    #    The cage is large (3)
    #    The cage is medium, sized (2)
    #    The cage is small (1)
    # We need an "if statement" to deal with each of these possibilities and we
    # need to update the large, small, and med counts along with the is_full 
    # parameter each time

    if cage_size == 3:

        large_count = large_count + 1

        # Now we need to take into account the possibility that the truck is
        # full of large boxes

        if large_count == 4:
            is_full = is_full + 1

    # Now we are back to 4 spaces for the next "if statement"

    if cage_size == 2: 
        med_count = med_count + 1

        if med_count == 6:
            is_full = is_full + 1

    if cage_size == 1:
        small_count = small_count +1

        if small_count == 12:

            is_full = is_full + 1

    # now we just need to return everything

    return is_full, large_count, med_count, small_count

# now for the main event!

# first, let's initialize the variables we need:

is_full = large_count = med_count = small_count = 0

# Now we need a "while loop" to drive all the "if statements". 
# In this while loop, we will be constantly redefining our variables, so we must
# create a list and set that equal to the new list being generate by the 
# functions above. You'll see how this is done in a second.

while is_full < 3:
    
    is_full, large_count, med_count, small_count = \
        organize_cats(is_full, large_count, med_count, small_count)

# See that slash at the end? Yeah, that means that the line continues.
# It's mainly a formatting thing.

# Once the while loop is finished, we can go from there.

print("The truck is full! Let's move out and leave ", large_count - 4,
      " large cats, ", med_count - 6," medium-sized, cats and ",
      small_count - 12," small cats behind! Who needs 'em?")

